Added Recycler - CEPH RBD

childsb opened the pull request:
Adding a CEPH RBD recycler.

childsb committed:
Added RBD Recycler and CLI Recycle override option.

childsb committed:
Merge remote-tracking branch 'upstream/master' into rbd_recycler

rootfs commented on pkg/volume/plugins.go:
```diff
@@ -411,12 +411,13 @@ func NewPersistentVolumeRecyclerPodTemplate() *api.Pod {
 					Image:   "gcr.io/google_containers/busybox",
 					Command: []string{"/bin/sh"},
-					Args:    []string{"-c", "test -e /scrub && echo $(date) > /scrub/trash.txt && find /scrub -mindepth 1 -maxdepth 1 -delete && test -z \"$(ls -A /scrub)\" || exit 1"},
+					Args:    []string{"-c", "test -e /scrub && rm -rf /scrub/..?* /scrub/.[!.]* /scrub/* ; echo $(date) > /scrub/trash.txt && exit 0 "},
```
what is the change for?

childsb replied:
it was broken here https://github.com/kubernetes/kubernetes/pull/16723#issuecomment-177784069

markturansky replied:
Probably want a separate PR for this change, since @thockin was backporting this fix to the older 1.1.x branch.  Can't do that with the fix being in this PR.

childsb replied:
Ok I created a new PR with just this fix here: https://github.com/kubernetes/kubernetes/pull/20495

markturansky replied:
I left a comment on #20495 about this scriptlet.  I understand this is needed to make the recycler work for RBD.  Better to remove it here, squash this PR, and then cherry pick the other commit from 20495.  When that merges, the cherry pick will fall out of this PR when you're hit with the rebase.

markturansky commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -150,7 +151,10 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.NamespaceSyncPeriod, "namespace-sync-period", s.NamespaceSyncPeriod, "The period for syncing namespace life-cycle updates")
 	fs.DurationVar(&s.PVClaimBinderSyncPeriod, "pvclaimbinder-sync-period", s.PVClaimBinderSyncPeriod, "The period for syncing persistent volumes and persistent volume claims")
 	fs.DurationVar(&s.MinResyncPeriod, "min-resync-period", s.MinResyncPeriod, "The resync period in reflectors will be random between MinResyncPeriod and 2*MinResyncPeriod")
+	
```
dead line

markturansky commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -150,7 +151,10 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
+	
 	fs.StringVar(&s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathNFS, "pv-recycler-pod-template-filepath-nfs", s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathNFS, "The file path to a pod definition used as a template for NFS persistent volume recycling")
+	fs.StringVar(&s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathRBD, "pv-recycler-pod-template-filepath-rbd", s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathRBD, "The file path to a pod definition used as a template for NFS persistent volume recycling")
+	
```
dead line

markturansky commented on cmd/kube-controller-manager/app/plugins.go:
```diff
@@ -72,6 +73,19 @@ func ProbeRecyclableVolumePlugins(flags options.VolumeConfigFlags) []volume.Volu
 	}
+
+	rbdConfig := volume.VolumeConfig{
+		RecyclerMinimumTimeout:   flags.PersistentVolumeRecyclerMinimumTimeoutNFS,
```
Good for testing and development to get the recycler working, but we'll want separate CLI flags for RBD.

childsb replied:
The current method of adding CLI flags for every recycler doesn't scale very well. I was hesitant to add new flags (creating API) that we may deprecate after 1.2

markturansky replied:
This and the increment require separate flags.  These are for NFS and we allow override per plugin.

markturansky commented on pkg/volume/rbd/rbd.go:
```diff
@@ -99,6 +102,8 @@ func (plugin *rbdPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, _ volume.Vo
 }
 
+
+
```
dead lines

markturansky commented:
Overall, looks pretty good.

Needs gofmt, too.

```
Verifying ./hack/../hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./cmd/kube-controller-manager/app/options/options.go
./cmd/kube-controller-manager/app/plugins.go
./pkg/volume/rbd/rbd.go
FAILED
```

markturansky commented:
@kubernetes/rh-storage

childsb committed:
Merge branch 'master' into rbd_recycler

childsb committed:
Fixed formatting.

erinboyd commented:
Update comment to be RBD rather than NFS to avoid confusion 
_`+	fs.StringVar(&s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathRBD, "pv-recycler-pod-template-filepath-rbd", s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathRBD, "The file path to a pod definition used as a template for NFS persistent vo`

erinboyd commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -150,7 +151,10 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
+
 	fs.StringVar(&s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathNFS, "pv-recycler-pod-template-filepath-nfs", s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathNFS, "The file path to a pod definition used as a template for NFS persistent volume recycling")
+	fs.StringVar(&s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathRBD, "pv-recycler-pod-template-filepath-rbd", s.VolumeConfigFlags.PersistentVolumeRecyclerPodTemplateFilePathRBD, "The file path to a pod definition used as a template for NFS persistent volume recycling")
+
```
Comment should be specific to RBD not NFS

childsb committed:
Change the options argument description for RBD.

childsb commented:
@markturansky @erinboyd I updated to reflect your comments.

markturansky commented:
@childsb are the correct commits pushed to GH?  It looks like the same source files without feedback.

markturansky commented:
From Travis: `Generated docs are out of date. Please run hack/update-generated-docs.sh`

childsb committed:
remove extra line breaks.

childsb committed:
Updated docs

childsb commented:
@markturansky I missed two line breaks which i fixed (thought the gofmt would pick it up.. sorry) and added the updated docs. Thanks for the feedback.

childsb committed:
Merge remote-tracking branch 'upstream/master' into rbd_recycler

markturansky commented on cmd/kube-controller-manager/app/plugins.go:
```diff
@@ -72,6 +73,19 @@ func ProbeRecyclableVolumePlugins(flags options.VolumeConfigFlags) []volume.Volu
+	if err := AttemptToLoadRecycler(flags.PersistentVolumeRecyclerPodTemplateFilePathRBD, &rbdConfig); err != nil {
+		glog.Fatalf("Could not create RBD recycler pod from file %s: %+v", flags.PersistentVolumeRecyclerPodTemplateFilePathRBD, err)
+	}
+
```
dead line

markturansky commented:
Couple more small issues.  Overall looking good.

@swagiaal 2nd review?

markturansky commented:
Also need a doc update to include RBD:  https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/persistent-volumes.md#recycling-policy

jsafrane commented on pkg/volume/rbd/rbd.go:
```diff
@@ -29,17 +29,20 @@ import (
-	exe  exec.Interface
+	host            volume.VolumeHost
+	exe             exec.Interface
+	newRecyclerFunc func(*volume.Spec, volume.VolumeHost, volume.VolumeConfig) (volume.Recycler, error)
```
@markturansky, why is `newRecyclerFunc` here? IMO it always leads to `newRecycler`. I would remove it from all the plugins with recycler (in a separate PR).

markturansky replied:
The idea is to allow a test to supply the factory function.  This used to be in more plugins w/ builder and cleaner but they've been refactored out.

I agree that this indirection (complexity) can be removed if it is not used anywhere.

jsafrane commented on pkg/volume/rbd/rbd.go:
```diff
@@ -235,3 +239,41 @@ func (plugin *rbdPlugin) execCommand(command string, args []string) ([]byte, err
+func (r *rbdRecycler) Recycle() error {
+	pod := r.config.RecyclerPodTemplate
+	pod.GenerateName = "pv-recycler-rbd-"
+	pod.Spec.Volumes[0].VolumeSource = api.VolumeSource{RBD: r.spec.PersistentVolume.Spec.RBD}
```
Using `pod.Spec.Volumes[0]` directly is quite dangerous, the array may be uninitialized.
The only place suggesting that there *must* be at least one item in `pod.Spec.Volumes` is here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L409

@markturansky, how are the admins supposed to know this when preparing the template? IMO we miss some documentation or a check in `AttemptToLoadRecycler()` (as separate PR, of course).

markturansky replied:
The code can be more defensive and ensure Volumes[0] exists.  That validation was added to the Origin version (https://github.com/openshift/origin/blob/master/pkg/cmd/server/kubernetes/master.go#L178).

Agree on docs and defensive coding for this.

jsafrane commented:
@childsb, you need to run hack/update-all.sh to regenerate docs for new cmdline options.

jsafrane commented:
LGTM, but we have some issues elsewhere in the recycler controller. I'll talk to @markturansky.

childsb committed:
Remove the fix for recycler scrub.  Will fix in another PR..

childsb commented:
Rebased and removed recycler fix
